File: slope_serial.py
#! /usr/bin/env python

import numpy as np

def main():

  input_file = open("aigrid.asc")
  params = input_file.readlines()
  input_file.close()
  ncols = params[0].split()[1]
  nrows = params[1].split()[1]
  xllcorner = params[2].split()[1]
  yllcorner = params[3].split()[1]
  cellsize = float(params[4].split()[1])
  NODATA = float(params[5].split()[1])

#raw_data = []
  #for line in params[6:]:
  #  raw_data.append(line.split())
  #data = np.asarray(raw_data)
  #data = data.astype(np.float)
  data = np.loadtxt(open("aigrid.asc"), skiprows=6)
  
  slope_data = calc_slope(data, cellsize, NODATA)
  #print slope_data
  #np.savetxt("output.txt", slope_data, fmt='%5.2f')

def calc_slope(grid, cellsize, NODATA):
  slope_grid = np.zeros_like(grid)
  for row in range(len(grid)):
    for col in range((len(grid[0]))):
      slope_grid[row][col] = cell_slope(grid, row, col, cellsize, NODATA)

  return slope_grid

def cell_slope(grid, row, col, cellsize, NODATA):
  #if row==len(grid)-1 or col==len(grid[0])-1 or row==0 or col==0:
    #TODO: Come back to this case later...
    #return NODATA

  #First, grab values for cells used in calculation
  nbhd = []
  for i in range(-1,2):
    for j in range(-1,2):
      #If out of bounds, log NODATA, these will be changed later.
      if row+i<=0 or row+i>=len(grid) or col+j<=0 or col+j>=len(grid[0]):
        nbhd.append(NODATA)
        continue
      nbhd.append(grid[row+i,col+j])

  if nbhd[4] == NODATA:
    return NODATA

  for i in range(len(nbhd)):
    if nbhd[i] == NODATA:
      nbhd[i] = nbhd[4]

  dz_dx = (nbhd[2] + 2*nbhd[5] + nbhd[8] - (nbhd[0] + 2*nbhd[3] + nbhd[6])) \
                / (8*cellsize)
  dz_dy = (nbhd[6] + 2*nbhd[7] + nbhd[8] - (nbhd[0] + 2*nbhd[1] + nbhd[2])) \
                / (8*cellsize)

  slope = np.arctan(np.sqrt(np.square(dz_dx) + np.square(dz_dy)))

  return slope

main()

File: slope_serial.cpp
/**************************************************
 * Note, just prints to stdout. Pipe to a file if desired.
 **************************************************/

#include <iostream>
#include <fstream>
#include <stdlib.h>
#include <string>
#include <sstream>
#include <math.h>
#include <algorithm>
#include <iterator>
#include <vector>

//static const int NCOLS = 1051;
//static const int NROWS = 1407;
//static const float CELLSIZE = 10.0;
//static const float NODATA = -32767.0;

int NCOLS, NROWS;
float CELLSIZE, NODATA;
string header[6];

using namespace std;

float** loadFile(string filename);
float** calcSlope(float** grid);
float cellSlope(float** grid, int row, int col);

int main(int argc, char* argv[]){
        if (argc != 2){
                cerr << "Incorrect number of args" << endl;
                cout << "Correct usage: ./slope_serial <filename>" << endl;
                return 1;
        }

        string filename(argv[1]);
        int i, j;

        //TODO: Free grid at end of program
        float** grid = loadFile(filename);
        float** slope_grid = calcSlope(grid);
        for (i=0; i<NROWS; i++){
                for (int j=0; j<NCOLS; j++){
                        cout << slope_grid[i][j] << " ";
                }
                cout << endl;
        }

        for (i=0; i<NROWS; i++){
                delete[] grid[i];
                delete[] slope_grid[i];
        }
        delete[] grid;
        delete[] slope_grid;

        return 0;
}

float** loadFile(string filename){
        //Open file
        ifstream inFile;
        inFile.open(filename.c_str());
        if(!inFile.is_open()){
                cerr << "File failed to open" << endl;
                exit(1);
        }
        
        int count = 0;
        vector<string> splits;
        
        while (count < 6)
	{
		//read in header line one at a time
		getline(inFile, header[count]);
		//grab number of cols and make into int variable
		if (count == 0)
		{
			istringstream stream(header[count]);
			copy(istream_iterator<string>(stream),
				istream_iterator<string>(),
				back_inserter(splits));
			NCOLS = stoi(splits[1], nullptr,10);
			count++;
		}
		//grab number of rows and make into int variable
		else if (count == 1)
		{
			istringstream stream(header[count]);
			copy(istream_iterator<string>(stream),
				istream_iterator<string>(),
				back_inserter(splits));
			NROWS = stoi(splits[3], nullptr, 10);
			count++;
		}
		//grab cell size and maake into float variable
		else if (count == 4)
		{
			istringstream stream(header[count]);
			copy(istream_iterator<string>(stream),
				istream_iterator<string>(),
				back_inserter(splits));
			CELLSIZE = stof(splits[5]);
			count++;
		}
		//grab no data value and make into float variable
		else if (count == 5)
		{
			istringstream stream(header[count]);
			copy(istream_iterator<string>(stream),
				istream_iterator<string>(),
				back_inserter(splits));
			NODATA = stof(splits[7]);
			count++;
		}
		//go to next line in header
		else
		count++;
	}
	
        float** grid = new float* [NROWS];
        for(int k=0; k<NROWS; k++){
                grid[k] = new float [NCOLS];
        }
        
        string line;
        
        int i=0;
        int j=0;
        //Read file line by line
        while(getline(inFile, line)){
                istringstream ss(line);
                string x;
                //Copy each token from a line into a grid
                while(getline(ss, x, ' ')){
                        grid[i][j] = atof(x.c_str());
                        j++;
                }
                j=0;
                i++;
        }
        inFile.close();
        return grid;
}

float** calcSlope(float** grid){
        float** slope_grid = new float* [NROWS];

        for (int row=0; row<NROWS; row++){
                slope_grid[row] = new float [NCOLS];
                for (int col=0; col<NCOLS; col++){
                        slope_grid[row][col] = cellSlope(grid, row, col);
                }
        }
        return slope_grid;
}

float cellSlope(float** grid, int row, int col){
        float nbhd[9];
        int k = 0;
        for (int i=-1; i<2; i++){
                for (int j=-1; j<2; j++){
                        if ((row+i<=0) || (row+i>=NROWS) || (col+j<=0) || (col+j>=NCOLS)){
                                nbhd[k] = NODATA;
                        }
                        else{
                                nbhd[k] = grid[row+i][col+j];
                        }
                        k++;
                }
        }

        if (nbhd[4] == NODATA){
                return nbhd[4];
        }

        for (k=0; k<9; k++){
                if (nbhd[k] == NODATA){
                        nbhd[k] = nbhd[4];
                }
        }

        float dz_dx = (nbhd[2] + (2*nbhd[5]) + nbhd[8] - (nbhd[0] + (2*nbhd[3]) + nbhd[6])) / (8*CELLSIZE);
        float dz_dy = (nbhd[6] + (2*nbhd[7]) + nbhd[8] - (nbhd[0] + (2*nbhd[1]) + nbhd[2])) / (8*CELLSIZE);

        return atan(sqrt(pow(dz_dx, 2) + pow(dz_dy, 2)));
}
